[optim] ASGD fold state updates into functional and pass list of vars rather than states

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70953
* #70484
* #70483
* #70482
* #70481
* #70295
* #70231
* #70230
* #70229
* #69982
* #69981
* #69980
* __->__ #71335
* #71334
* #71333

Differential Revision: [D33767871](https://our.internmc.facebook.com/intern/diff/D33767871)

mikaylagawarecki committed:
[optim] NAdam fold state updates into functional and pass list of vars rather than states

[ghstack-poisoned]

albanD commented on torch/optim/_multi_tensor/_functional.py:
```diff
@@ -162,28 +166,32 @@ def asgd(params: List[Tensor],
-        state['mu'] = 1 / max(1, state['step'] - t0)
+    for i in range(len(mus)):
+        new_eta = torch.tensor(lr / math.pow((1 + lambd * lr * state_steps[i].item()), alpha))
+        etas[i].copy_(new_eta)
```
you can do `etas[i].fill_(XXX)` to avoid creating the tensor and the copy_

mikaylagawarecki committed:
Update on "[optim] ASGD fold state updates into functional and pass list of vars rather than states"




[ghstack-poisoned]

albanD reviewed (approved):
Sounds good!

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767871).

mikaylagawarecki committed:
Update on "[optim] ASGD fold state updates into functional and pass list of vars rather than states"


Differential Revision: [D33767871](https://our.internmc.facebook.com/intern/diff/D33767871)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767871).
